### PR TITLE
feat(form-field): add support for separate label and placeholder

### LINKED ISSUE #6194: feat(input, select): separate the placeholder and label as shown in the official mocks
#### Bug, feature request, or proposal:
Bug

#### What is the expected behavior?
The placeholder used in a textarea with multiple rows should automatically use up the other rows when too long to fit in just the first row.

#### What is the current behavior?
The placeholder is truncated in the first row with the other rows left blank.

#### What are the steps to reproduce?
Plunk: http://plnkr.co/edit/OAxiFX?p=preview

#### What is the use-case or motivation for changing an existing behavior?
I would want to tell the user as to what format a textarea is to be filled up. Possibly referencing certain standard examples / policies for which I need a few lines' worth of text.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular 4.3.1
Material 2.0.0-beta.8-374aaff
All OS
Typescript ^2.2.2
Checked in Chrome & Firefox

#### Is there anything else we should know?
No

### PR DESCRIPTION
**Pull request by crisbeto**
* Adds the ability for the user to specify both a label and placeholder text for a form field.
* Renames all of the "floating placeholder" related terminology to refer to a "floating label" instead.
* Aligns the input placeholder color with the Material spec.

Fixes #6194.

### CODE REVIEW

**Comment by crisbeto** [comment]
@mmalerba I'm open to using the native `label` instead of `mat-label`, but it would require use to have a directive that selects all `label` nodes.

**Comment by jelbourn** [comment]
Could this be done without any breaking changes?

**Comment by crisbeto** [comment]
The only real breaking change is renaming the `shouldPlaceholderFloat` to `shouldLabelFloat` in the `MatFormFieldControl` interface @jelbourn, everything else is still re-exported under its old name. I can re-add the `shouldPlaceholderFloat` as an optional value and mark it as deprecated, but I suspect that not many people are implementing the interface.

**Comment by jelbourn** [comment]
Let's go with that- I want to anything we have to call a breaking API change

**Comment by crisbeto** [comment]
Refactored to avoid breaking changes.

**Comment by mmalerba** [comment]
Still need to take a closer look at the code, but had a couple thoughts after looking at the demo:
1. It should be possible to have in input with no label and a placeholder, but making that work now would be a breaking change. Maybe we can just file a new issue and update that when we're making breaking changes
2. The placeholder needs an animation so it doesn't overlap with the label when the label is animating up. The spec has a [video](https://storage.googleapis.com/material-design/publish/material_v_12/assets/0B8wSqcLwbhFuREg4Q0R4M3VGejA/textfields-presslight-01.webm) showing what it should look like

Thanks for doing this! Something I've been wanting to do for a while but haven't gotten the time

**Comment by crisbeto** [comment]
1. Yeah I wanted to avoid breaking changes.
2. We can't really animate the native placeholder attribute, the best we could do is delay showing it until the animation is done, but that means switching over the placeholder animations to Angular animations. We could do it in a follow-up.

**Comment by mmalerba** [comment]
we might not be able to do the sequenced animation, but could we at least slap a transition on the opacity? I think it will make it look a little nicer

**Comment by crisbeto** [comment]
We could do that, but it would be a progressive enhancement since not all browsers allow you to transition the native placeholder (e.g. Chrome does, but Firefox doesn't).

**Comment by crisbeto** [comment]
Rebased and made it so the placeholder doesn't overlap the animating label @mmalerba, can you take another look?

File: src/lib/form-field/form-field.html [hunk 1]
```diff
@@ -8,24 +8,31 @@
     <div class="mat-input-infix mat-form-field-infix">
       <ng-content></ng-content>
 
-      <span class="mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper">
+      <span class="mat-form-field-label-wrapper">
```
Comment (mmalerba) [inline src/lib/form-field/form-field.html]: I'm sure people are targeting those classes with custom styles, lets leave them for now with a TODO to remove them at the next major.

File: src/lib/form-field/form-field.html [hunk 2]
```diff
@@ -8,24 +8,31 @@
...
         <!-- We add aria-owns as a workaround for an issue in JAWS & NVDA where the label isn't
              read if it comes before the control in the DOM. -->
-        <label class="mat-input-placeholder mat-form-field-placeholder"
+        <label class="mat-form-field-label"
```
Comment (mmalerba) [inline src/lib/form-field/form-field.html]: Same, leave the old classes too for now

File: src/lib/form-field/form-field.ts [hunk 3]
```diff
@@ -220,24 +227,40 @@ export class MatFormField implements AfterViewInit, AfterContentInit, AfterConte
...
   _hasPlaceholder() {
     return !!(this._control.placeholder || this._placeholderChild);
   }
 
+  _hasLabel() {
+    return !!this._labelChild;
+  }
+
+  _shouldLabelFloat() {
+    return this._canLabelFloat && (this._control.shouldLabelFloat ||
```
Comment (mmalerba) [inline src/lib/form-field/form-field.ts]: This logic seems different than before, it didn't consider `_canPlaceholderFloat` before

File: src/lib/form-field/form-field.ts [hunk 4]
```diff
@@ -220,24 +227,40 @@ export class MatFormField implements AfterViewInit, AfterContentInit, AfterConte
...
+        this._control.shouldPlaceholderFloat || this._shouldAlwaysFloat);
+  }
+
+  _hideControlPlaceholder() {
+    return !this._hasLabel() || !this._shouldLabelFloat();
```
Comment (mmalerba) [inline src/lib/form-field/form-field.ts]: Maybe the `_canLabelFloat` from above should be factored into this logic instead?

File: src/lib/form-field/form-field.ts (hunk 3 above)
Comment (crisbeto) [reply src/lib/form-field/form-field.ts]: If I remember correctly, some of the logic was done by combining CSS classes. I simplified it to make it easier to follow.

File: src/lib/form-field/form-field.ts (hunk 4 above)
Comment (crisbeto) [reply src/lib/form-field/form-field.ts]: The `_shouldLabelFloat` call here already factors in `_canLabelFloat`.

**Comment by crisbeto** [comment]
I restored the classes and addressed the comments @mmalerba, can you take another look?

**Comment by jelbourn** [comment]
@crisbeto ready to presubmit if you can rebase

**Comment by crisbeto** [comment]
Rebased @jelbourn.

**Comment by tinayuangao** [comment]
@crisbeto please rebase. Thanks!

**Comment by crisbeto** [comment]
Rebased @tinayuangao.

**Pull request merged**